### PR TITLE
add platform target properties to compiler.

### PR DESCRIPTION
**Pull request by vankop**
**What kind of change does this PR introduce?**

add 2 methods `compiler.getPlatformTargetInfo` and `compiler.setPlatformTargetInfo` to allow get and set platform target properties. by default resolve this properties from target option.

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
`compiler.getPlatformTargetInfo` and `compiler.setPlatformTargetInfo`. Also new `PlatformPlugin` that allow to change platform options, e.g. `new PlatformPlugin({ node: true })`

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
Looks good, but let's apply this things for loader context too, I don't want to use hacky `_compiler` in loaders to get these things, for example css-loader can export only locals when you your target is node

**Comment by vankop** [comment]
could we add this in another PR? @alexander-akait

File: lib/config/defaults.js [hunk 1]
```diff
@@ -315,6 +325,16 @@ const applyWebpackOptionsDefaults = options => {
...
+
+	if (targetProperties === false) return false;
+	return {
+		web: targetProperties.web,
+		browser: targetProperties.browser,
+		webworker: targetProperties.webworker,
+		node: targetProperties.node,
+		nwjs: targetProperties.nwjs,
+		electron: targetProperties.electron
+	};
```
[the remark this answers is not in the document]
Comment (vankop) [reply lib/config/defaults.js:337]: rest of the target properties related to environment, they are accessible in `output.environment`

[the remark this answers is not in the document]
Comment (vankop) [reply lib/config/defaults.js:337]: https://github.com/webpack/webpack/blob/31ddbd223225ded0118b2bce62e09f6059390027/lib/config/defaults.js#L1099

**Comment by alexander-akait** [comment]
@vankop Let's rebase

**Review by alexander-akait — APPROVED** [review]

**Pull request merged**